E0716: confusing reference to borrow of `Stdin` value when there's no obvious reference within `StdinLock`

Given the following code: [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4129b8aca043f4e6a3c6ea755c5ed954)

```rust
use std::io::{self, BufRead};
fn main() {
    let locked = io::stdin().lock();
    for line in locked.lines() {
        println!("{}", line.unwrap());
    }
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error[E0716]: temporary value dropped while borrowed
 --> src/main.rs:3:18
  |
3 |     let locked = io::stdin().lock();
  |                  ^^^^^^^^^^^       - temporary value is freed at the end of this statement
  |                  |
  |                  creates a temporary which is freed while still in use
4 |     for line in locked.lines() {
  |                 ------ borrow later used here
  |
  = note: consider using a `let` binding to create a longer lived value

error: aborting due to previous error

For more information about this error, try `rustc --explain E0716`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

I think it's typical for a newcomer to want to get a `Lines` iterator by chaining method calls like `io::stdin().lock().lines()`. There's no obvious reason why that shouldn't work. The examples in the `std::io` documentation do typically show `io::stdin()` being assigned to a variable prior to calling its `.lock()` method, but don't explain why that's necessary.

I think the "borrow" terminology is misleading, but there might not be a concise way to describe what's actually going on to the user. The "temporary value is freed at the end of this statement" is probably correct. The "creates a temporary which is freed while still in use" is probably more correctly "creates a temporary that is required to outlive (or live exactly as long as) the result of...", followed by a pointer to the `.lock()` method invocation, with text of "...this method call". Maybe it should also point out the `'_` lifetime argument in the `StdinLock<'_>` return type.

I tried and failed to find a satisfactory explanation of what exactly `StdinLock` was borrowing from the temporary `Stdin` value produced by `io::stdin()`. When I looked through the library source code, the only references in `StdinLock` were those created by locking a `Mutex`, and the `Mutex` in question is static. I'm guessing the existence of a borrow is assumed rather than inferred from analyzing the function body of `io::Stdin::lock()`. The assumption might be the result of the anonymous lifetime argument on the output type `StdinLock<'_>` from `io::Stdin::lock()`, which imposes a requirement that the `StdinLock` doesn't outlive (or has the same lifetime as?) the `Stdin` value that is the receiver of `lock()`.

The correct fix might be to change the locked stdin handles to all use `'static` lifetimes, but I will probably file that as a separate issue.

As a simplified example: [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=84f6bb1b93633e45af71e8eab9fc44e9)

```rust
static NUM: i32 = 1;
struct A;
struct B<'a>(&'a i32);
impl A {
    fn get(&self) -> B<'_> {
        // Contains NO ref to self whatsoever!
        B(&NUM)
    }
}
fn a() -> A { A }
fn main() {
    let b = a().get();
    println!("{}", b.0.to_string());
}
```

```
error[E0716]: temporary value dropped while borrowed
  --> src/main.rs:12:13
   |
12 |     let b = a().get();
   |             ^^^      - temporary value is freed at the end of this statement
   |             |
   |             creates a temporary which is freed while still in use
13 |     println!("{}", b.0.to_string());
   |                    --- borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value
```

The error message refers to a borrow even when the output type of `get()` does not refer to `self` at all!

Shorter example, probably better suited for an example in the compiler error index: [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a4b633db02a4b043f4a057b7ab5d56b1)

```rust
fn main() {
    static NUM: i32 = 17;
    fn foo() -> i32 { 42 }
    // output ref is totally unrelated to input
    fn bar(_: &i32) -> &i32 { &NUM }
    let x = bar(&foo());
    println!("{}", x);
}
```

```
error[E0716]: temporary value dropped while borrowed
 --> src/main.rs:6:18
  |
6 |     let x = bar(&foo());
  |                  ^^^^^ - temporary value is freed at the end of this statement
  |                  |
  |                  creates a temporary which is freed while still in use
7 |     println!("{}", x);
  |                    - borrow later used here
  |
  = note: consider using a `let` binding to create a longer lived value
```

--- a/library/std/src/io/mod.rs
+++ b/library/std/src/io/mod.rs
@@ -277,8 +277,12 @@ pub use self::error::{Error, ErrorKind, Result};
 pub use self::stdio::set_output_capture;
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use self::stdio::{stderr, stdin, stdout, Stderr, Stdin, Stdout};
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub use self::stdio::{stderr_locked, stdin_locked, stdout_locked};
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use self::stdio::{StderrLock, StdinLock, StdoutLock};
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub use self::stdio::{StderrOwnedLock, StdinOwnedLock, StdoutOwnedLock};
 #[unstable(feature = "print_internals", issue = "none")]
 pub use self::stdio::{_eprint, _print};
 #[stable(feature = "rust1", since = "1.0.0")]

--- a/library/std/src/io/mod.rs
+++ b/library/std/src/io/mod.rs
@@ -281,8 +281,6 @@ pub use self::stdio::{stderr, stdin, stdout, Stderr, Stdin, Stdout};
 pub use self::stdio::{stderr_locked, stdin_locked, stdout_locked};
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use self::stdio::{StderrLock, StdinLock, StdoutLock};
-#[unstable(feature = "stdio_locked", issue = "none")]
-pub use self::stdio::{StderrOwnedLock, StdinOwnedLock, StdoutOwnedLock};
 #[unstable(feature = "print_internals", issue = "none")]
 pub use self::stdio::{_eprint, _print};
 #[stable(feature = "rust1", since = "1.0.0")]

--- a/library/std/src/io/stdio.rs
+++ b/library/std/src/io/stdio.rs
@@ -261,6 +261,21 @@ pub struct StdinLock<'a> {
     inner: MutexGuard<'a, BufReader<StdinRaw>>,
 }
 
+/// Owned locked [`Stdin`] handle, returned by [`Stdin::into_lock`] and
+/// [`io::stdin_locked`].
+///
+/// This is exactly like [`StdinLock`], except that it can outlive the
+/// [`Stdin`] handle that was used to create it. See the [`StdinLock`]
+/// documentation for more details.
+///
+/// ### Note: Windows Portability Consideration
+///
+/// When operating in a console, the Windows implementation of this stream does not support
+/// non-UTF-8 byte sequences. Attempting to read bytes that are not valid UTF-8 will return
+/// an error.
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub type StdinOwnedLock = StdinLock<'static>;
+
 /// Constructs a new handle to the standard input of the current process.
 ///
 /// Each handle returned is a reference to a shared global buffer whose access
@@ -310,6 +325,48 @@ pub fn stdin() -> Stdin {
     }
 }
 
+/// Constructs a new locked handle to the standard input of the current
+/// process.
+///
+/// Each handle returned is a guard granting locked access to a shared
+/// global buffer whose access is synchronized via a mutex. If you need
+/// more explicit control over locking, for example, in a multi-threaded
+/// program, use the [`io::stdin`] function to obtain an unlocked handle,
+/// along with the [`Stdin::lock`] method.
+///
+/// The lock is released when the returned guard goes out of scope. The
+/// returned guard also implements the [`Read`] and [`BufRead`] traits for
+/// accessing the underlying data.
+///
+/// **Note**: The mutex locked by this handle is not reentrant. Even in a
+/// single-threaded program, calling other code that accesses [`Stdin`]
+/// could cause a deadlock or panic, if this locked handle is held across
+/// that call.
+///
+/// ### Note: Windows Portability Consideration
+/// When operating in a console, the Windows implementation of this stream does not support
+/// non-UTF-8 byte sequences. Attempting to read bytes that are not valid UTF-8 will return
+/// an error.
+///
+/// # Examples
+///
+/// ```no_run
+/// #![feature(stdio_locked)]
+/// use std::io::{self, Read};
+///
+/// fn main() -> io::Result<()> {
+///     let mut buffer = String::new();
+///     let mut handle = io::stdin_locked();
+///
+///     handle.read_to_string(&mut buffer)?;
+///     Ok(())
+/// }
+/// ```
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub fn stdin_locked() -> StdinOwnedLock {
+    stdin().into_lock()
+}
+
 impl Stdin {
     /// Locks this handle to the standard input stream, returning a readable
     /// guard.
@@ -334,7 +391,7 @@ impl Stdin {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn lock(&self) -> StdinLock<'_> {
-        StdinLock { inner: self.inner.lock().unwrap_or_else(|e| e.into_inner()) }
+        self.lock_any()
     }
 
     /// Locks this handle and reads a line of input, appending it to the specified buffer.
@@ -366,6 +423,43 @@ impl Stdin {
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn read_line(&self, buf: &mut String) -> io::Result<usize> {
         self.lock().read_line(buf)
+    }
+
+    // Locks this handle with any lifetime. This depends on the
+    // implementation detail that the underlying `Mutex` is static.
+    fn lock_any<'a>(&self) -> StdinLock<'a> {
+        StdinLock { inner: self.inner.lock().unwrap_or_else(|e| e.into_inner()) }
+    }
+
+    /// Consumes this handle to the standard input stream, locking the
+    /// shared global buffer associated with the stream and returning a
+    /// readable guard.
+    ///
+    /// The lock is released when the returned guard goes out of scope. The
+    /// returned guard also implements the [`Read`] and [`BufRead`] traits
+    /// for accessing the underlying data.
+    ///
+    /// It is often simpler to directly get a locked handle using the
+    /// [`stdin_locked`] function instead, unless nearby code also needs to
+    /// use an unlocked handle.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// #![feature(stdio_locked)]
+    /// use std::io::{self, Read};
+    ///
+    /// fn main() -> io::Result<()> {
+    ///     let mut buffer = String::new();
+    ///     let mut handle = io::stdin().into_lock();
+    ///
+    ///     handle.read_to_string(&mut buffer)?;
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "stdio_locked", issue = "none")]
+    pub fn into_lock(self) -> StdinOwnedLock {
+        self.lock_any()
     }
 }
 
@@ -507,6 +601,20 @@ pub struct StdoutLock<'a> {
     inner: ReentrantMutexGuard<'a, RefCell<LineWriter<StdoutRaw>>>,
 }
 
+/// Owned locked [`Stdout`] handle, returned by [`Stdout::into_lock`] and
+/// [`io::stdout_locked`].
+///
+/// This is exactly like [`StdoutLock`], except that it can outlive the
+/// [`Stdout`] handle that was used to create it. See the [`StdoutLock`]
+/// documentation for more details.
+///
+/// ### Note: Windows Portability Consideration
+/// When operating in a console, the Windows implementation of this stream does not support
+/// non-UTF-8 byte sequences. Attempting to write bytes that are not valid UTF-8 will return
+/// an error.
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub type StdoutOwnedLock = StdoutLock<'static>;
+
 static STDOUT: SyncOnceCell<ReentrantMutex<RefCell<LineWriter<StdoutRaw>>>> = SyncOnceCell::new();
 
 /// Constructs a new handle to the standard output of the current process.
@@ -558,6 +666,42 @@ pub fn stdout() -> Stdout {
     }
 }
 
+/// Constructs a new locked handle to the standard output of the current
+/// process.
+///
+/// Each handle returned is a guard granting locked access to a shared
+/// global buffer whose access is synchronized via a mutex. If you need
+/// more explicit control over locking, for example, in a multi-threaded
+/// program, use the [`io::stdout`] function to obtain an unlocked handle,
+/// along with the [`Stdout::lock`] method.
+///
+/// The lock is released when the returned guard goes out of scope. The
+/// returned guard also implements the [`Write`] trait for writing data.
+///
+/// ### Note: Windows Portability Consideration
+/// When operating in a console, the Windows implementation of this stream does not support
+/// non-UTF-8 byte sequences. Attempting to write bytes that are not valid UTF-8 will return
+/// an error.
+///
+/// # Examples
+///
+/// ```no_run
+/// #![feature(stdio_locked)]
+/// use std::io::{self, Write};
+///
+/// fn main() -> io::Result<()> {
+///     let mut handle = io::stdout_locked();
+///
+///     handle.write_all(b"hello world")?;
+///
+///     Ok(())
+/// }
+/// ```
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub fn stdout_locked() -> StdoutLock<'static> {
+    stdout().into_lock()
+}
+
 pub fn cleanup() {
     if let Some(instance) = STDOUT.get() {
         // Flush the data and disable buffering during shutdown
@@ -595,7 +739,44 @@ impl Stdout {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn lock(&self) -> StdoutLock<'_> {
+        self.lock_any()
+    }
+
+    // Locks this handle with any lifetime. This depends on the
+    // implementation detail that the underlying `ReentrantMutex` is
+    // static.
+    fn lock_any<'a>(&self) -> StdoutLock<'a> {
         StdoutLock { inner: self.inner.lock() }
+    }
+
+    /// Consumes this handle to the standard output stream, locking the
+    /// shared global buffer associated with the stream and returning a
+    /// writable guard.
+    ///
+    /// The lock is released when the returned lock goes out of scope. The
+    /// returned guard also implements the [`Write`] trait for writing data.
+    ///
+    /// It is often simpler to directly get a locked handle using the
+    /// [`io::stdout_locked`] function instead, unless nearby code also
+    /// needs to use an unlocked handle.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// #![feature(stdio_locked)]
+    /// use std::io::{self, Write};
+    ///
+    /// fn main() -> io::Result<()> {
+    ///     let mut handle = io::stdout().into_lock();
+    ///
+    ///     handle.write_all(b"hello world")?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "stdio_locked", issue = "none")]
+    pub fn into_lock(self) -> StdoutOwnedLock {
+        self.lock_any()
     }
 }
 
@@ -717,6 +898,20 @@ pub struct StderrLock<'a> {
     inner: ReentrantMutexGuard<'a, RefCell<StderrRaw>>,
 }
 
+/// Owned locked [`Stderr`] handle, returned by [`Stderr::into_lock`] and
+/// [`io::stderr_locked`].
+///
+/// This is exactly like [`StderrLock`], except that it can outlive the the
+/// [`Stderr`] handle that was used to create it. See the [`StderrLock`]
+/// documentation for more details.
+///
+/// ### Note: Windows Portability Consideration
+/// When operating in a console, the Windows implementation of this stream does not support
+/// non-UTF-8 byte sequences. Attempting to write bytes that are not valid UTF-8 will return
+/// an error.
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub type StderrOwnedLock = StderrLock<'static>;
+
 /// Constructs a new handle to the standard error of the current process.
 ///
 /// This handle is not buffered.
@@ -769,6 +964,35 @@ pub fn stderr() -> Stderr {
     }
 }
 
+/// Constructs a new locked handle to the standard error of the current
+/// process.
+///
+/// This handle is not buffered.
+///
+/// ### Note: Windows Portability Consideration
+/// When operating in a console, the Windows implementation of this stream does not support
+/// non-UTF-8 byte sequences. Attempting to write bytes that are not valid UTF-8 will return
+/// an error.
+///
+/// # Example
+///
+/// ```no_run
+/// #![feature(stdio_locked)]
+/// use std::io::{self, Write};
+///
+/// fn main() -> io::Result<()> {
+///     let mut handle = io::stderr_locked();
+///
+///     handle.write_all(b"hello world")?;
+///
+///     Ok(())
+/// }
+/// ```
+#[unstable(feature = "stdio_locked", issue = "none")]
+pub fn stderr_locked() -> StderrOwnedLock {
+    stderr().into_lock()
+}
+
 impl Stderr {
     /// Locks this handle to the standard error stream, returning a writable
     /// guard.
@@ -792,7 +1016,41 @@ impl Stderr {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn lock(&self) -> StderrLock<'_> {
+        self.lock_any()
+    }
+
+    // Locks this handle with any lifetime. This depends on the
+    // implementation detail that the underlying `ReentrantMutex` is
+    // static.
+    fn lock_any<'a>(&self) -> StderrLock<'a> {
         StderrLock { inner: self.inner.lock() }
+    }
+
+    /// Locks and consumes this handle to the standard error stream,
+    /// returning a writable guard.
+    ///
+    /// The lock is released when the returned guard goes out of scope. The
+    /// returned guard also implements the [`Write`] trait for writing
+    /// data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(stdio_locked)]
+    /// use std::io::{self, Write};
+    ///
+    /// fn foo() -> io::Result<()> {
+    ///     let stderr = io::stderr();
+    ///     let mut handle = stderr.into_lock();
+    ///
+    ///     handle.write_all(b"hello world")?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "stdio_locked", issue = "none")]
+    pub fn into_lock(self) -> StderrOwnedLock {
+        self.lock_any()
     }
 }
 

--- a/library/std/src/io/stdio.rs
+++ b/library/std/src/io/stdio.rs
@@ -261,21 +261,6 @@ pub struct StdinLock<'a> {
     inner: MutexGuard<'a, BufReader<StdinRaw>>,
 }
 
-/// Owned locked [`Stdin`] handle, returned by [`Stdin::into_lock`] and
-/// [`io::stdin_locked`].
-///
-/// This is exactly like [`StdinLock`], except that it can outlive the
-/// [`Stdin`] handle that was used to create it. See the [`StdinLock`]
-/// documentation for more details.
-///
-/// ### Note: Windows Portability Consideration
-///
-/// When operating in a console, the Windows implementation of this stream does not support
-/// non-UTF-8 byte sequences. Attempting to read bytes that are not valid UTF-8 will return
-/// an error.
-#[unstable(feature = "stdio_locked", issue = "none")]
-pub type StdinOwnedLock = StdinLock<'static>;
-
 /// Constructs a new handle to the standard input of the current process.
 ///
 /// Each handle returned is a reference to a shared global buffer whose access
@@ -363,8 +348,8 @@ pub fn stdin() -> Stdin {
 /// }
 /// ```
 #[unstable(feature = "stdio_locked", issue = "none")]
-pub fn stdin_locked() -> StdinOwnedLock {
-    stdin().into_lock()
+pub fn stdin_locked() -> StdinLock<'static> {
+    stdin().into_locked()
 }
 
 impl Stdin {
@@ -451,14 +436,14 @@ impl Stdin {
     ///
     /// fn main() -> io::Result<()> {
     ///     let mut buffer = String::new();
-    ///     let mut handle = io::stdin().into_lock();
+    ///     let mut handle = io::stdin().into_locked();
     ///
     ///     handle.read_to_string(&mut buffer)?;
     ///     Ok(())
     /// }
     /// ```
     #[unstable(feature = "stdio_locked", issue = "none")]
-    pub fn into_lock(self) -> StdinOwnedLock {
+    pub fn into_locked(self) -> StdinLock<'static> {
         self.lock_any()
     }
 }
@@ -601,20 +586,6 @@ pub struct StdoutLock<'a> {
     inner: ReentrantMutexGuard<'a, RefCell<LineWriter<StdoutRaw>>>,
 }
 
-/// Owned locked [`Stdout`] handle, returned by [`Stdout::into_lock`] and
-/// [`io::stdout_locked`].
-///
-/// This is exactly like [`StdoutLock`], except that it can outlive the
-/// [`Stdout`] handle that was used to create it. See the [`StdoutLock`]
-/// documentation for more details.
-///
-/// ### Note: Windows Portability Consideration
-/// When operating in a console, the Windows implementation of this stream does not support
-/// non-UTF-8 byte sequences. Attempting to write bytes that are not valid UTF-8 will return
-/// an error.
-#[unstable(feature = "stdio_locked", issue = "none")]
-pub type StdoutOwnedLock = StdoutLock<'static>;
-
 static STDOUT: SyncOnceCell<ReentrantMutex<RefCell<LineWriter<StdoutRaw>>>> = SyncOnceCell::new();
 
 /// Constructs a new handle to the standard output of the current process.
@@ -699,7 +670,7 @@ pub fn stdout() -> Stdout {
 /// ```
 #[unstable(feature = "stdio_locked", issue = "none")]
 pub fn stdout_locked() -> StdoutLock<'static> {
-    stdout().into_lock()
+    stdout().into_locked()
 }
 
 pub fn cleanup() {
@@ -767,7 +738,7 @@ impl Stdout {
     /// use std::io::{self, Write};
     ///
     /// fn main() -> io::Result<()> {
-    ///     let mut handle = io::stdout().into_lock();
+    ///     let mut handle = io::stdout().into_locked();
     ///
     ///     handle.write_all(b"hello world")?;
     ///
@@ -775,7 +746,7 @@ impl Stdout {
     /// }
     /// ```
     #[unstable(feature = "stdio_locked", issue = "none")]
-    pub fn into_lock(self) -> StdoutOwnedLock {
+    pub fn into_locked(self) -> StdoutLock<'static> {
         self.lock_any()
     }
 }
@@ -898,20 +869,6 @@ pub struct StderrLock<'a> {
     inner: ReentrantMutexGuard<'a, RefCell<StderrRaw>>,
 }
 
-/// Owned locked [`Stderr`] handle, returned by [`Stderr::into_lock`] and
-/// [`io::stderr_locked`].
-///
-/// This is exactly like [`StderrLock`], except that it can outlive the the
-/// [`Stderr`] handle that was used to create it. See the [`StderrLock`]
-/// documentation for more details.
-///
-/// ### Note: Windows Portability Consideration
-/// When operating in a console, the Windows implementation of this stream does not support
-/// non-UTF-8 byte sequences. Attempting to write bytes that are not valid UTF-8 will return
-/// an error.
-#[unstable(feature = "stdio_locked", issue = "none")]
-pub type StderrOwnedLock = StderrLock<'static>;
-
 /// Constructs a new handle to the standard error of the current process.
 ///
 /// This handle is not buffered.
@@ -989,8 +946,8 @@ pub fn stderr() -> Stderr {
 /// }
 /// ```
 #[unstable(feature = "stdio_locked", issue = "none")]
-pub fn stderr_locked() -> StderrOwnedLock {
-    stderr().into_lock()
+pub fn stderr_locked() -> StderrLock<'static> {
+    stderr().into_locked()
 }
 
 impl Stderr {
@@ -1041,7 +998,7 @@ impl Stderr {
     ///
     /// fn foo() -> io::Result<()> {
     ///     let stderr = io::stderr();
-    ///     let mut handle = stderr.into_lock();
+    ///     let mut handle = stderr.into_locked();
     ///
     ///     handle.write_all(b"hello world")?;
     ///
@@ -1049,7 +1006,7 @@ impl Stderr {
     /// }
     /// ```
     #[unstable(feature = "stdio_locked", issue = "none")]
-    pub fn into_lock(self) -> StderrOwnedLock {
+    pub fn into_locked(self) -> StderrLock<'static> {
         self.lock_any()
     }
 }

--- a/library/std/src/io/stdio/tests.rs
+++ b/library/std/src/io/stdio/tests.rs
@@ -1,5 +1,6 @@
 use super::*;
 use crate::panic::{RefUnwindSafe, UnwindSafe};
+use crate::sync::mpsc::sync_channel;
 use crate::thread;
 
 #[test]
@@ -44,4 +45,137 @@ fn panic_doesnt_poison() {
     let _a = _a.lock();
     let _a = stderr();
     let _a = _a.lock();
+}
+
+#[test]
+fn stderr_owned_lock_static() {
+    assert_static::<StderrOwnedLock>();
+}
+#[test]
+fn stdin_owned_lock_static() {
+    assert_static::<StdinOwnedLock>();
+}
+#[test]
+fn stdout_owned_lock_static() {
+    assert_static::<StdoutOwnedLock>();
+}
+
+fn assert_static<T: 'static>() {}
+
+#[test]
+#[cfg_attr(target_os = "emscripten", ignore)]
+fn test_lock_stderr() {
+    test_lock(stderr, stderr_locked);
+}
+#[test]
+#[cfg_attr(target_os = "emscripten", ignore)]
+fn test_lock_stdin() {
+    test_lock(stdin, stdin_locked);
+}
+#[test]
+#[cfg_attr(target_os = "emscripten", ignore)]
+fn test_lock_stdout() {
+    test_lock(stdout, stdout_locked);
+}
+
+// Helper trait to make lock testing function generic.
+trait Stdio<'a>: 'static
+where
+    Self::Lock: 'a,
+{
+    type Lock;
+    fn lock(&'a self) -> Self::Lock;
+}
+impl<'a> Stdio<'a> for Stderr {
+    type Lock = StderrLock<'a>;
+    fn lock(&'a self) -> StderrLock<'a> {
+        self.lock()
+    }
+}
+impl<'a> Stdio<'a> for Stdin {
+    type Lock = StdinLock<'a>;
+    fn lock(&'a self) -> StdinLock<'a> {
+        self.lock()
+    }
+}
+impl<'a> Stdio<'a> for Stdout {
+    type Lock = StdoutLock<'a>;
+    fn lock(&'a self) -> StdoutLock<'a> {
+        self.lock()
+    }
+}
+
+// Helper trait to make lock testing function generic.
+trait StdioOwnedLock: 'static {}
+impl StdioOwnedLock for StderrOwnedLock {}
+impl StdioOwnedLock for StdinOwnedLock {}
+impl StdioOwnedLock for StdoutOwnedLock {}
+
+// Tests locking on stdio handles by starting two threads and checking that
+// they block each other appropriately.
+fn test_lock<T, U>(get_handle: fn() -> T, get_locked: fn() -> U)
+where
+    T: for<'a> Stdio<'a>,
+    U: StdioOwnedLock,
+{
+    // State enum to track different phases of the test, primarily when
+    // each lock is acquired and released.
+    #[derive(Debug, PartialEq)]
+    enum State {
+        Start1,
+        Acquire1,
+        Start2,
+        Release1,
+        Acquire2,
+        Release2,
+    }
+    use State::*;
+    // Logging vector to be checked to make sure lock acquisitions and
+    // releases happened in the correct order.
+    let log = Arc::new(Mutex::new(Vec::new()));
+    let ((tx1, rx1), (tx2, rx2)) = (sync_channel(0), sync_channel(0));
+    let th1 = {
+        let (log, tx) = (Arc::clone(&log), tx1);
+        thread::spawn(move || {
+            log.lock().unwrap().push(Start1);
+            let handle = get_handle();
+            {
+                let locked = handle.lock();
+                log.lock().unwrap().push(Acquire1);
+                tx.send(Acquire1).unwrap(); // notify of acquisition
+                tx.send(Release1).unwrap(); // wait for release command
+                log.lock().unwrap().push(Release1);
+            }
+            tx.send(Acquire1).unwrap(); // wait for th2 acquire
+            {
+                let locked = handle.lock();
+                log.lock().unwrap().push(Acquire1);
+            }
+            log.lock().unwrap().push(Release1);
+        })
+    };
+    let th2 = {
+        let (log, tx) = (Arc::clone(&log), tx2);
+        thread::spawn(move || {
+            tx.send(Start2).unwrap(); // wait for start command
+            let locked = get_locked();
+            log.lock().unwrap().push(Acquire2);
+            tx.send(Acquire2).unwrap(); // notify of acquisition
+            tx.send(Release2).unwrap(); // wait for release command
+            log.lock().unwrap().push(Release2);
+        })
+    };
+    assert_eq!(rx1.recv().unwrap(), Acquire1); // wait for th1 acquire
+    log.lock().unwrap().push(Start2);
+    assert_eq!(rx2.recv().unwrap(), Start2); // block th2
+    assert_eq!(rx1.recv().unwrap(), Release1); // release th1
+    assert_eq!(rx2.recv().unwrap(), Acquire2); // wait for th2 acquire
+    assert_eq!(rx1.recv().unwrap(), Acquire1); // block th1
+    assert_eq!(rx2.recv().unwrap(), Release2); // release th2
+    th2.join().unwrap();
+    th1.join().unwrap();
+    assert_eq!(
+        *log.lock().unwrap(),
+        [Start1, Acquire1, Start2, Release1, Acquire2, Release2, Acquire1, Release1]
+    );
 }

--- a/library/std/src/io/stdio/tests.rs
+++ b/library/std/src/io/stdio/tests.rs
@@ -48,21 +48,6 @@ fn panic_doesnt_poison() {
 }
 
 #[test]
-fn stderr_owned_lock_static() {
-    assert_static::<StderrOwnedLock>();
-}
-#[test]
-fn stdin_owned_lock_static() {
-    assert_static::<StdinOwnedLock>();
-}
-#[test]
-fn stdout_owned_lock_static() {
-    assert_static::<StdoutOwnedLock>();
-}
-
-fn assert_static<T: 'static>() {}
-
-#[test]
 #[cfg_attr(target_os = "emscripten", ignore)]
 fn test_lock_stderr() {
     test_lock(stderr, stderr_locked);
@@ -107,9 +92,9 @@ impl<'a> Stdio<'a> for Stdout {
 
 // Helper trait to make lock testing function generic.
 trait StdioOwnedLock: 'static {}
-impl StdioOwnedLock for StderrOwnedLock {}
-impl StdioOwnedLock for StdinOwnedLock {}
-impl StdioOwnedLock for StdoutOwnedLock {}
+impl StdioOwnedLock for StderrLock<'static> {}
+impl StdioOwnedLock for StdinLock<'static> {}
+impl StdioOwnedLock for StdoutLock<'static> {}
 
 // Tests locking on stdio handles by starting two threads and checking that
 // they block each other appropriately.

